Additions for evolutions 3

nilsdeppe opened the pull request:
## Proposed changes

- Adds a type alias for the Jacobian type, similar to the `InverseJacobian` type alias
- Fixes a bug in `compute_item_result_impl` that caused really confusing template errors
- Adds tags for the logical coordinates (item) and inverse jacobian (compute item)
- Generalizes the spatial derivative prefix `d` to be a compute item and item depending on its template parameters
- Renames `Tags::d` to `Tags::deriv`
- Fixes a clang-tidy warning in DataBox.hpp

These are again changes I had to make for getting evolutions working.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Add type alias for Jacobian

nilsdeppe committed:
Change partial derivative tag from 'd' to 'deriv'

nilsdeppe committed:
Add Tags for LogicalCoordinates and InverseJacobian

The LogicalCoordinates is an item, while the InversJacobian is a
compute item.

nilsdeppe committed:
Fix clang-tidy warning in DataBox.hpp

kidder commented on src/DataStructures/Tensor/TypeAliases.hpp:
```diff
@@ -334,11 +334,15 @@ using ijaa = Tensor<DataType, tmpl::integral_list<std::int32_t, 3, 2, 1, 1>,
            typelist<SpatialIndex<Dim, UpLo::Up, Frame1>,
                     SpatialIndex<Dim, UpLo::Lo, Frame2>>>;
+
+template <size_t Dim, typename Frame1, typename Frame2>
```
I think we should use SourceFrame and TargetFrame as the template parameters (and above in InverseJacobian) to match CoordinateMap

nilsdeppe replied:
I agree, `Frame1` -> `SourceFrame` and `Frame2` -> `TargetFrame`?

kidder commented on src/DataStructures/DataBoxTag.hpp:
```diff
@@ -414,7 +423,7 @@ struct is_variables_compute_item<
           typename... Args>
 using wrap_tags_in =
-    tmpl::transform<TagList, tmpl::bind<Wrapper, tmpl::_1, Args...>>;
+    tmpl::transform<TagList, tmpl::bind<Wrapper, tmpl::_1, tmpl::pin<Args>...>>;
```
is this related to the commit message?

nilsdeppe replied:
Indirectly, yes. I'll split it into a separate commit

kidder commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,122 @@
+ * parameters:
+ * 1. The tag to wrap
+ * 2. The volume dim as a type (e.g. `tmpl::size_t<Dim>`)
+ * 3. The frame of the derivative index is in `Tensor`s
```
index is in -> index in

kidder commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,122 @@
+ * `Tags::Variables<tmpl::list<Tags...>>`, the list of tags to be
+ * differentiated, and the Tag for inverse Jacobian between the logical frame
+ * and the frame in which the derivative is taken. Note that the derivative Tags
+ * must be a contiguous subset from the first tag in the `Tags...` pack.
```
from the head of the `Tags...` pack

kidder commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,122 @@
+template <class>
+class Variables;
+
+template <typename DerivativeTags, typename VariableTags, size_t Dim>
```
Having to include these forward declarations suggests to me that we might want to move the compute item tags to where the function is.  Thoughts?

nilsdeppe replied:
I think there are sort of two options for other locations:

1. A tag file closer to the function, but we still forward declare the functions. This will reduce annoying compilation dependencies that will otherwise crop up. That is, if you only need the Tag but don't need the function definition you shouldn't need the function's header (for function templates only is this a concern)

2. We add the Tag to the header of the function. For functions where we have explicit instantiations and the header file does not contain a definition this is probably ideal. For function templates where we can't generate explicit instantiations this could be less optimal.

Have said all that, only functions that act on Variables will we not be able to generate explicit instantiations trivially. However, there aren't too many of these, and what we can do is have the definition in a cpp file, that cpp file then gets included in a cpp file for the system, which generates the explicit instantiations and all problems are solved. 2 + generating instantiations for specific systems by putting the responsibility on that system's author is best.

kidder replied:
I think I like the last suggestion best; but I'm happy to hear other opinions

wthrowe replied:
I agree, I like 2.

nilsdeppe committed:
Fix compute_item_result_impl

nilsdeppe committed:
Add missing 'pin' to wrap_tags_in metafunction

nilsdeppe commented:
I've addressed Larry's requests. The outstanding one from his is what to do with compute item tags, which I left a long reply to.

wthrowe commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,123 @@
+class DataVector;
+template <size_t Dim>
+class Index;
+template <class>
```
typename

wthrowe commented on src/Domain/Tags.hpp:
```diff
@@ -16,7 +16,27 @@
+/// \ingroup ComputationalDomainGroup
+/// The logical coordinates in the Element
+template <size_t Dim>
+struct LogicalCoordinates : db::DataBoxTag {
```
Can this be a compute item based on the extents?

nilsdeppe replied:
Yep! Good point!

wthrowe commented on src/Domain/Tags.hpp:
```diff
@@ -46,6 +66,23 @@ struct ElementMap : db::DataBoxTag {
+/// the map.
+template <typename MapTag, typename SourceCoordsTag>
+struct InverseJacobian : db::ComputeItemTag {
+  static constexpr db::DataBoxString label = "InverseJacobian";
```
Can this be a prefix tag so we don't get a name conflict once we have multiple maps?

wthrowe commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp:
```diff
@@ -368,3 +371,162 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearOperators.PartialDerivs",
+  const auto& du = db::get<deriv_tag>(box);
+
+  for (size_t n = 0; n < du.size(); ++n) {
+    CAPTURE_PRECISE(du.data()[n] - expected_du.data()[n]);  // NOLINT
```
Explain NOLINT

nilsdeppe replied:
@kidder why is this NOLINT? I stole it from what you wrote :stuck_out_tongue:

kidder replied:
pointer arithmetic

wthrowe commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,123 @@
+               Tags::Extents<derivative_frame_index::dim>, InverseJacobianTag>;
+};
+
+// Logical partial derivatives
```
Do we have a known use for derivatives in the logical coordinates?  I expected them to be used to calculate derivatives in other frames, but it looks like we're not doing that.

wthrowe commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,123 @@
+struct deriv<tmpl::list<VariablesTags...>, tmpl::list<DerivativeTags...>,
+             InverseJacobianTag,
+             Requires<tt::is_a_v<Tensor, db::item_type<InverseJacobianTag>>>>
+    : db::ComputeItemTag {
```
This looks like it will give us label conflicts as soon as we have multiple Variables or multiple frames.  I don't know what to do about it here, though, since the prefix tag stuff only uses one parameter.

wthrowe commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -14,32 +16,123 @@
   static constexpr db::DataBoxString label = "d";
 };
 
+// Partial derivatives with derivative index in frame related to by
```
related to...the logical coordinates, maybe?

nilsdeppe committed:
Remove Prefixes.hpp dependency from Test_DataBox

nilsdeppe committed:
Move deriv tag to PartialDerivatives, and add cpp file for impls

We are moving towards having tags for compute items be in the header
corresponding to the function it wraps. This moves the partial
derivative tag to PartialDerivatives.hpp and enables it as a compute
item. The implementation of partial derivatives is also moved into a
PartialDerivatives.cpp, even though they take Variables classes. The
Variables classes will need to be explicitly instantiated for each
system. For example, the instantiations could be at the end of the
Equations.cpp by including the PartialDerivatives.cpp and then adding
the instantiations.

nilsdeppe committed:
Simplify includes for GeneralizedHarmonic

kidder commented on src/Domain/Tags.hpp:
```diff
@@ -32,10 +33,14 @@ namespace Tags {
-  using type = tnsr::I<DataVector, Dim, Frame::Logical>;
+  using argument_tags = tmpl::list<ExtentsTag>;
+  static constexpr auto function = logical_coordinates<ExtentsTag::dim>;
+  // (const Index<Dim>& extents) noexcept {
```
remove

kidder commented on src/Domain/Tags.hpp:
```diff
@@ -32,10 +33,14 @@ namespace Tags {
 /// The logical coordinates in the Element
-template <size_t Dim>
-struct LogicalCoordinates : db::DataBoxTag {
+template <typename ExtentsTag>
```
hard-code Extents

kidder commented on src/DataStructures/DataBox/Prefixes.hpp:
```diff
@@ -15,31 +15,6 @@
 #include "Utilities/TypeTraits.hpp"
 
 namespace Tags {
-/// \ingroup DataBoxTagsGroup
```
couple of typos in commit message move -> moved and at the end of the

kidder commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp:
```diff
@@ -0,0 +1,183 @@
+struct LogicalImpl;
+}  // namespace partial_derivatives_detail
+
+/// \ingroup NumericalAlgorithmsGroup
```
remove documentation

kidder commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -6,177 +6,134 @@
+ * 2. The volume dim as a type (e.g. `tmpl::size_t<Dim>`)
+ * 3. The frame of the derivative index in `Tensor`s
+ *
+ * The partial derivatives in a Frame that is not the logical frame as a compute
```
The third variant

kidder commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -6,177 +6,134 @@
+ * and the frame in which the derivative is taken. Note that the derivative Tags
+ * must be a contiguous subset from the head of the `Tags...` pack.
+ *
+ * The partial derivatives in the logical frame can also be computed. In that
```
The fourth (or final) variant
